### DISCUSSION TITLE
disable Link component

### DISCUSSION THREAD

**Discussion by quant-daddy** [vercel/next.js#21396] [Help]
Is there a way to disable the Link component? I thought if `href` is `#`, it would disable the Link component. However, very surprisingly, the actual link in this becomes the `pathname` of the current page and it leads on a runtime error on click the link.

**Comment by lednhatkhanh** [comment]
I also want to disable Next.js link but still show the url.

**Comment by skfahad** [comment]
I passed empty object in `href` attribute of Link component then on click of child `anchor` element called the `preventDefault()` method as follows

```
<Link href={conditionMet ? '/some-path' : {}}>
    <a onClick={(event) => !conditionMet ? event.preventDefault() : null}>
        Some Link
    </a>
</Link>
```

**Reply by xyfll7** [reply]
``` tsx
<Link href={"/test"} scroll={false} onClick={(e) => { e.preventDefault()  }}>
                test
</Link>
```

**Reply by OFD16** [reply]
you are amayzing bro thanks a lot @xyfll7

**Reply by vikaskrsharma** [reply]
Not sure the seo side effects but this definitely solved all the problems.

**Reply by hossein-deyri** [reply]
this is good

**Comment by saurabhburade** [comment]
You can create a wrapper component for the Link to disable it conditionally like following
```
const LinkWrapper = ({ href, ...props }) => {
  if (href == "#") {
    return <button {...props}>Click Here</button>;
  }
  return (
    <Link href={href} {...props}>
      <button>Click Here</button>
    </Link>
  );
};
```

**Reply by devlopersabbir** [reply]
this is a good way???

**Comment by camposdev** [comment]
I just changed the `as` prop to `div`, so the `href` does not work in a `div` or another element like `button`:
```
<Link href={href} as={isDisabled ? 'div' : 'a'} />
```

**Reply by romoore** [reply]
This just changed the browser bar URL, it didn't change the component type.

**Reply by devlopersabbir** [reply]
`as` work as like href 🐸

**Comment by sahilsian** [comment]
This suggestion is a little bit hacky, and only works in some scenarios but I just change styling if a value is passed through ``href``. Just to make it seem like links are not clickable, and if they do they are just sent to the top of the page.

```jsx
<Link style={{
    textDecoration: href ? "underline" : "none",
    cursor: href ? "pointer" : "default"
}}  href={href ? href : {}}>
    <Text_Component
        font_size={font_size}
        line_height={line_height}
        font_weight={font_weight}
        color={color}
        text={text}
        align={align}
        hover={href}
    >
    </Text_Component>
</Link>
```

 Of course this is not the best way and the best way would be to return a different element if ``href`` is not passed. But that was already mentioned by @saurabhburade comment, and this also works fine for me.

**Comment by CapitaineToinon** [comment]
Yeah I wrote something like this for a shadcn button / link that can be disabled

```ts
type PaginationButtonProps =
	| (ComponentProps<typeof Link> & { disabled?: false })
	| (ComponentPropsWithoutRef<'button'> & { disabled: true })

function PaginationButton(props: PaginationButtonProps) {
	if (props.disabled) {
		const { disabled, ...rest } = props
		return (
			<Button disabled={disabled} asChild>
				<button {...rest} />
			</Button>
		)
	}

	return (
		<Button asChild>
			<Link {...props} />
		</Button>
	)
}
```

Used like this for example:

```ts
<ButtonGroup>
	<PaginationButton
		href={`?page=${page - 1}`}
		disabled={isFirstPage}
	>
		Previous
	</PaginationButton>
	<PaginationButton
		href={`?page=${page + 1}`}
		disabled={isLastPage}
	>
		Next
	</PaginationButton>
</ButtonGroup>
```

**Reply by devlopersabbir** [reply]
Thanks you so much, this is the one of the best solutions :)

**Reply by owencraston** [reply]
This is it 👍

**Comment by akomm** [comment]
Using css:
```css
a.disabled {
  pointer-events: none;
}
```
Then component:
```tsx
<Link className={clsx(p.disabled && "disabled")} aria-disabled={p.disabled}>...</Link>
```

Alt using tailwind utilities: 
```tsx
<Link className={clsx(p.disabled && "pointer-events-none")} aria-disabled={p.disabled}>...</Link>
```

Alt. using style:
```tsx
<Link style={{pointerEvents: "none"}} aria-disabled={p.disabled}>...</Link>
```

`aria-disabled` is for accessibility.

**Reply by FedericoMainettiGambera** [reply]
To also disable keyboard navigation: `tabIndex={disabled ? -1 : undefined}`

**Reply by akomm** [reply]
good point

**Reply by tayejoseph** [reply]
using `pointer-events: none` on the css end will prevent `cursor: no-drop` from working

**Reply by akomm** [reply]
Yes. Then you need to use another solution or wrap the element in another element that has a cursor set. 
May I add thought that using cursor as an indicator for that is either redundant, if the element has already a clear formatting depicting it as disabled or in case it is not otherwise depicted as disabled the disabled state isn't accessible/depictable on devices without a cursor.

So typically you should not do that, but I get that maybe you have some older app that still uses this technique or you have someone who wants it and does not want to talk about this matter, then you can only use JS solution or a wrapper element (which you can render only when disabled). But that is also not very nice for accessibility without extra effort.

**Comment by yigithancolak** [comment]
Update: 
In my situation i used the existing pathname with  usePathname()  and change the pointer events and opacity according to it. Simply if i am already in that route, link will have a disable effect with tailwind classes.
```js
<Link
              key={title}
              href={route}
              className={`btn-primary flex items-center ${
                pathname === route && 'pointer-events-none opacity-20'
              }`}
            >
              <span className='hidden md:inline mr-1'>{title}</span>
              <button.icon className='inline' fontSize={20} />
            </Link>
```

**Reply by CapitaineToinon** [reply]
You should do that if you care about your links behaving normally. You solution disables next's prefetch on hover, the native bahvior of opening links in a new tab my middle clicking without the scroll wheel and more.

**Reply by yigithancolak** [reply]
Yes you re right. Wish they had direct way to do it but links are not treated as button element, they are more like a element so some extra effort needed for sure.

**Comment by angelitolm** [comment]
My solution:

```
interface LinkWrapperProps {
  href: string;
  disabled?: boolean;
  children?: any;
  passHref?: boolean;
  className?: string;
}
```

```
const LinkWrapper = ({ children, href, disabled, passHref, ...props }: LinkWrapperProps) => {
  if (disabled)
    return <Button disabled={disabled} {...props}>{children}</Button>;

  return <Link href={href} passHref={passHref} {...props}>{children}</Link>;
};
```

Use mode:

```
<LinkWrapper href={href} passHref className='flex gap-1' disabled={!isAvailable(href)}>
  {icon}
  <Trans i18nKey={`navigation:tabs.${label}`} components={{ i: <i />, br: <br /> }} />
</LinkWrapper>
```

**Comment by vasilemidrigan** [comment]
> Is there a way to disable the Link component? I thought if `href` is `#`, it would disable the Link component. However, very surprisingly, the actual link in this becomes the `pathname` of the current page and it leads on a runtime error on click the link.

Maybe as a solution: 

```
<button disabled>
    <Link key={v4()} href={`?page=${pageIndex}`}>
        {pageIndex}
    </Link>
</button>
```

**Reply by akomm** [reply]
accessibility

**Comment by MeroCallys** [comment]
This works for me.
```
 <Link
                              href={condition ? `/orders/${order.id}` : "#"}
                              className={clsx(
                                " hover:text-gray-400",
                                {
                                  "text-gray-400 pointer-events-none":
                                    !disabled[1] ? true : false,
                                },
                                {
                                  "text-gray-500": condition ? true : false,
                                }
                              )}
                            >
                              <ArrowUpRightSquare size={20} />
                            </Link>
                          </>
```

**Comment by Hammadhammad2** [comment]
we can use e.Preventdefault() and can apply condition to diable the link on onClicking event.
 <Link
                className=' absolute -bottom-[1.875rem] right-0 whitespace-nowrap md:-bottom-[1.375rem] 2xl:-bottom-[1.75rem]'
                to={'/forgot-password'}
                onClick={(e) => {
                  if (loading) e.preventDefault();
                }}
              >
                Forgot Password?
              </Link>

**Comment by austinschrader** [comment]
Why is this so hard

**Reply by amanzrx4** [reply]
just nextjs things

**Comment by jtmuller5** [comment]
You can use Tailwind to conditionally ignore pointer event:

```
<Link
          href={`/booking/preferences?zip=${zip}`}
          className={`${disabled ? 'pointer-events-none': ''}`}
        >
          <button className="w-full">Search</button>
</Link>
```

**Reply by akomm** [reply]
https://github.com/vercel/next.js/discussions/21396#discussioncomment-6362701

**Comment by KhanhLe1306** [comment]
Try **pointerEvents** , it disables the link
```
 <Link
       href={`/yourLink`}
       style={{
              textDecoration: 'none',
              pointerEvents: yourCondition ? 'auto' : 'none'
                }}
       key={...}
       target="_blank"
>
       <OtherComponent />
</Link>
```

**Reply by akomm** [reply]
The answer is just above yours linked ;P
https://github.com/vercel/next.js/discussions/21396#discussioncomment-6362701

**Comment by derb3n** [comment]
<Button disabled={disabled} asChild={!disabled}>
        <Link href={'/home'}>Home</Link>
      </Button>

works for me with the shadcn button
